// @shopify/cli/dist/cli/services/store/auth/result.js
import { outputCompleted, outputInfo, outputResult, outputToken, outputContent } from '@shopify/cli-kit/node/output';
function serializeStoreAuthResult(result) {
    return JSON.stringify(result, null, 2);
}
function buildStoreAuthSuccessText(result) {
    const displayName = result.associatedUser?.email ? ` as ${result.associatedUser.email}` : '';
    return {
        completed: ['Logged in.', `Authenticated${displayName} against ${result.store}.`],
        info: ['', 'To verify that authentication worked, run:', `shopify store execute --store ${result.store} --query 'query { shop { name id } }'`],
    };
}
function displayStoreAuthOpeningBrowser() {
    outputInfo('Shopify CLI will open the app authorization page in your browser.');
    outputInfo('');
}
function displayStoreAuthManualAuthUrl(authorizationUrl) {
    outputInfo('Browser did not open automatically. Open this URL manually:');
    outputInfo(outputContent `${outputToken.link(authorizationUrl)}`);
    outputInfo('');
}
function displayStoreAuthResult(result, format = 'text') {
    if (format === 'json') {
        outputResult(serializeStoreAuthResult(result));
        return;
    }
    const text = buildStoreAuthSuccessText(result);
    text.completed.forEach((line) => outputCompleted(line));
    text.info.forEach((line) => outputInfo(line));
}
export function createStoreAuthPresenter(format = 'text') {
    return {
        openingBrowser: displayStoreAuthOpeningBrowser,
        manualAuthUrl: displayStoreAuthManualAuthUrl,
        success(result) {
            displayStoreAuthResult(result, format);
        },
    };
}
