// @shopify/cli/dist/cli/services/store/auth/token-client.js
import { adminUrl } from '@shopify/cli-kit/node/api/admin';
import { graphqlRequest } from '@shopify/cli-kit/node/api/graphql';
import { AbortError } from '@shopify/cli-kit/node/error';
import { fetch } from '@shopify/cli-kit/node/http';
import { outputContent, outputDebug, outputToken } from '@shopify/cli-kit/node/output';
import { maskToken, STORE_AUTH_APP_CLIENT_ID } from './config.js';
function truncateHttpErrorBody(body, length = 300) {
    return body.slice(0, length);
}
export async function exchangeStoreAuthCodeForToken(options) {
    const endpoint = `https://${options.store}/admin/oauth/access_token`;
    outputDebug(outputContent `Exchanging authorization code for token at ${outputToken.raw(endpoint)}`);
    const response = await fetch(endpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            client_id: STORE_AUTH_APP_CLIENT_ID,
            code: options.code,
            code_verifier: options.codeVerifier,
            redirect_uri: options.redirectUri,
        }),
    });
    const body = await response.text();
    if (!response.ok) {
        outputDebug(outputContent `Token exchange failed with HTTP ${outputToken.raw(String(response.status))}: ${outputToken.raw(truncateHttpErrorBody(body || response.statusText))}`);
        throw new AbortError(`Failed to exchange OAuth code for an access token (HTTP ${response.status}).`, body || response.statusText);
    }
    let parsed;
    try {
        parsed = JSON.parse(body);
    }
    catch {
        throw new AbortError('Received an invalid token response from Shopify.');
    }
    outputDebug(outputContent `Token exchange succeeded: access_token=${outputToken.raw(maskToken(parsed.access_token))}, refresh_token=${outputToken.raw(parsed.refresh_token ? maskToken(parsed.refresh_token) : 'none')}, expires_in=${outputToken.raw(String(parsed.expires_in ?? 'unknown'))}s, user=${outputToken.raw(String(parsed.associated_user?.id ?? 'unknown'))} (${outputToken.raw(parsed.associated_user?.email ?? 'no email')})`);
    return parsed;
}
export async function refreshStoreAccessToken(options) {
    const endpoint = `https://${options.store}/admin/oauth/access_token`;
    outputDebug(outputContent `Refreshing access token for ${outputToken.raw(options.store)} using refresh_token=${outputToken.raw(maskToken(options.refreshToken))}`);
    const response = await fetch(endpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            client_id: STORE_AUTH_APP_CLIENT_ID,
            grant_type: 'refresh_token',
            refresh_token: options.refreshToken,
        }),
    });
    const body = await response.text();
    if (!response.ok) {
        outputDebug(outputContent `Token refresh failed with HTTP ${outputToken.raw(String(response.status))}: ${outputToken.raw(truncateHttpErrorBody(body || response.statusText))}`);
        throw new AbortError(`Token refresh failed for ${options.store} (HTTP ${response.status}).`);
    }
    let parsed;
    try {
        parsed = JSON.parse(body);
    }
    catch {
        throw new AbortError('Received an invalid refresh response from Shopify.');
    }
    if (!parsed.access_token) {
        throw new AbortError(`Token refresh returned an invalid response for ${options.store}.`);
    }
    return {
        accessToken: parsed.access_token,
        refreshToken: parsed.refresh_token,
        expiresIn: parsed.expires_in,
        refreshTokenExpiresIn: parsed.refresh_token_expires_in,
    };
}
const CurrentAppInstallationAccessScopesQuery = `#graphql
  query CurrentAppInstallationAccessScopes {
    currentAppInstallation {
      accessScopes {
        handle
      }
    }
  }
`;
export async function fetchCurrentStoreAuthScopes(options) {
    outputDebug(outputContent `Fetching current app installation scopes for ${outputToken.raw(options.store)} using token ${outputToken.raw(maskToken(options.accessToken))}`);
    const data = await graphqlRequest({
        query: CurrentAppInstallationAccessScopesQuery,
        api: 'Admin',
        url: adminUrl(options.store, 'unstable'),
        token: options.accessToken,
        responseOptions: { handleErrors: false },
    });
    if (!Array.isArray(data.currentAppInstallation?.accessScopes)) {
        throw new Error('Shopify did not return currentAppInstallation.accessScopes.');
    }
    return data.currentAppInstallation.accessScopes.flatMap((scope) => typeof scope.handle === 'string' ? [scope.handle] : []);
}
