// @shopify/mini-oxygen/dist/vite/plugin.js
import { defaultClientConditions } from "vite";
import {
  createMiniOxygenDevEnvironment,
  mergeMiniOxygenRuntimeOptions
} from "./environment.js";
import {
  setupOxygenMiddleware
} from "./server-middleware.js";
const DEFAULT_SSR_ENTRY = "./server";
function oxygen(pluginOptions = {}) {
  let apiOptions = {};
  let miniOxygenEnvironment;
  const resolveMiniOxygenOptions = async (runtimeOptions, viteDevServer) => {
    const entry = runtimeOptions.entry ?? pluginOptions.entry ?? DEFAULT_SSR_ENTRY;
    const remoteEnv = await Promise.resolve(runtimeOptions.envPromise);
    return {
      entry,
      viteDevServer,
      crossBoundarySetup: runtimeOptions.crossBoundarySetup,
      env: { ...remoteEnv, ...runtimeOptions.env, ...pluginOptions.env },
      debug: runtimeOptions.debug ?? pluginOptions.debug ?? false,
      inspectorPort: runtimeOptions.inspectorPort ?? pluginOptions.inspectorPort,
      requestHook: runtimeOptions.requestHook,
      entryPointErrorHandler: runtimeOptions.entryPointErrorHandler,
      compatibilityDate: runtimeOptions.compatibilityDate,
      logRequestLine: (
        // Give priority to the plugin option over the CLI option here,
        // since the CLI one is just a default, not a user-provided flag.
        pluginOptions.logRequestLine ?? runtimeOptions.logRequestLine
      )
    };
  };
  const applyRuntimeOptions = (newOptions) => {
    miniOxygenEnvironment?.configureRuntime(newOptions);
    apiOptions = mergeMiniOxygenRuntimeOptions(apiOptions, newOptions);
  };
  return [
    {
      name: "oxygen:main",
      config(config, env) {
        return {
          appType: "custom",
          resolve: {
            conditions: ["worker", "workerd", ...defaultClientConditions]
          },
          ssr: {
            noExternal: true,
            target: "webworker",
            resolve: {
              conditions: ["worker", "workerd", ...defaultClientConditions]
            }
          },
          // When building, the CLI will set the `ssr` option to `true`
          // if no --entry flag is passed for the default SSR entry file.
          // Replace it here with a default value.
          ...env.isSsrBuild && config.build?.ssr && {
            build: {
              ssr: config.build?.ssr === true ? (
                // No --entry flag passed by the user, use the
                // option passed to the plugin or the default value
                pluginOptions.entry ?? DEFAULT_SSR_ENTRY
              ) : (
                // --entry flag passed by the user, keep it
                config.build?.ssr
              )
            }
          }
        };
      },
      configEnvironment(name) {
        if (name !== "ssr") return;
        return {
          resolve: {
            conditions: ["worker", "workerd", ...defaultClientConditions]
          },
          dev: {
            createEnvironment(name2, config) {
              return miniOxygenEnvironment = createMiniOxygenDevEnvironment(
                name2,
                config,
                apiOptions,
                resolveMiniOxygenOptions
              );
            }
          }
        };
      },
      api: {
        registerPluginOptions(newOptions) {
          applyRuntimeOptions(newOptions);
        }
      },
      configureServer: {
        order: "pre",
        handler: (viteDevServer) => {
          return () => {
            setupOxygenMiddleware(
              viteDevServer,
              () => apiOptions.entryPointErrorHandler
            );
          };
        }
      },
      generateBundle() {
        if (apiOptions.compatibilityDate) {
          if (!/^\d{4}-\d{2}-\d{2}$/.test(apiOptions.compatibilityDate)) {
            throw new Error(
              `Invalid compatibility date "${apiOptions.compatibilityDate}"`
            );
          }
          const oxygenJsonFile = "oxygen.json";
          const oxygenJsonContent = {
            version: 1,
            compatibility_date: apiOptions.compatibilityDate
          };
          this.emitFile({
            type: "asset",
            fileName: oxygenJsonFile,
            source: JSON.stringify(oxygenJsonContent, null, 2)
          });
        }
      }
    }
  ];
}
export {
  oxygen
};
